### PR TITLE
release: 0.8.1

### LINKED ISSUE #220: setup: upgrading to `adage==0.10.2` breaks yadage workflows
After upgrading to`adage==0.10.2`, it's required to upgrade `networkx>=2.4`. This makes our yadage workflows fail on the workflow engine side. The jobs finished successfully but there's a problem afterward, perhaps when the visualization files are generated.

```console
$ reana-client logs -w hello-yadage.1
==> Workflow engine logs
Workflow exited unexpectedly.
Command 'convert -delay 50 $(ls /var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/*.png|sort) /var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/workflow.gif' returned non-zero exit status 1.
2022-02-02 10:08:28,350 | yadage.creators | MainThread | INFO | initializing workflow with initdata: {'helloworld': 'code/helloworld.py', 'inputfile': 'data/names.txt', 'sleeptime': 0} discover: True relative: True
2022-02-02 10:08:28,350 | adage.pollingexec | MainThread | INFO | preparing adage coroutine.
2022-02-02 10:08:28,350 | adage | MainThread | INFO | starting state loop.
2022-02-02 10:08:28,387 | yadage.wflowview | MainThread | INFO | added </init:0|defined|unknown>
2022-02-02 10:08:28,562 | yadage.wflowview | MainThread | INFO | added </helloworld:0|defined|unknown>
2022-02-02 10:08:28,655 | adage.pollingexec | MainThread | INFO | submitting nodes [</init:0|defined|known>]
2022-02-02 10:08:28,709 |       pack.init.step |   INFO | publishing data: <TypedLeafs: {'helloworld': '/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/code/helloworld.py', 'inputfile': '/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/data/names.txt', 'sleeptime': 0}>
2022-02-02 10:08:28,709 | adage | MainThread | INFO | unsubmittable: 0 | submitted: 0 | successful: 0 | failed: 0 | total: 2 | open rules: 0 | applied rules: 2
2022-02-02 10:08:43,824 | adage.node | MainThread | INFO | node ready </init:0|success|known>
2022-02-02 10:08:43,824 | adage.pollingexec | MainThread | INFO | submitting nodes [</helloworld:0|defined|known>]
2022-02-02 10:08:44,196 | reana-workflow-engine-yadage | MainThread | INFO | Submitted job with id: ec9c6a38-5da1-4abf-92f2-1d51c8884aae
2022-02-02 10:08:44,196 | adage | MainThread | INFO | unsubmittable: 0 | submitted: 0 | successful: 1 | failed: 0 | total: 2 | open rules: 0 | applied rules: 2
2022-02-02 10:08:59,379 | adage.node | MainThread | INFO | node ready </helloworld:0|success|known>
2022-02-02 10:08:59,379 | adage | MainThread | INFO | unsubmittable: 0 | submitted: 0 | successful: 2 | failed: 0 | total: 2 | open rules: 0 | applied rules: 2
2022-02-02 10:09:14,398 | adage.controllerutils | MainThread | INFO | no nodes can be run anymore and no rules are applicable
2022-02-02 10:09:14,399 | adage.controllerutils | MainThread | INFO | no nodes can be run anymore and no rules are applicable
2022-02-02 10:09:14,399 | adage | MainThread | INFO | unsubmittable: 0 | submitted: 0 | successful: 2 | failed: 0 | total: 2 | open rules: 0 | applied rules: 2
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
Error: <stdin>: syntax error in line 1 near '{'
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_00.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_01.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_02.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_03.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_04.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_05.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_06.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_07.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_08.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_09.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_10.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_11.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_12.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_13.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_14.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_15.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_16.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_17.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_18.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_19.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_20.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_21.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_22.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_23.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_24.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_25.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_26.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_27.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_28.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_29.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_30.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_31.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_32.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_33.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_34.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_35.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_36.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_37.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_38.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_39.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: improper image header `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/track/dag_40.png' @ error/png.c/ReadPNGImage/4092.
convert-im6.q16: no images defined `/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/_yadage/adage/workflow.gif' @ error/convert.c/ConvertImageCommand/3258.
2022-02-02 10:09:15,113 | yadage.steering_api | MainThread | INFO | done. dumping workflow to disk.
Couldn't import dot_parser, loading of dot files will not be possible.




==> Job logs
==> Step: helloworld
==> Workflow ID: 1bb4b9de-3d77-452e-8d59-c49be4086293
==> Compute backend: Kubernetes
==> Job ID: reana-run-job-d300724b-021e-4893-b276-8a18026766d5
==> Docker image: python:2.7-slim
==> Command: python "/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/code/helloworld.py" --sleeptime 0 --inputfile "/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/data/names.txt" --outputfile "/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/helloworld/greetings.txt"
==> Status: finished
==> Finished: 2022-02-02T10:08:59
==> Logs:
job: :
 2022-02-02T10:08:44 INFO     Parameters: inputfile=/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/data/names.txt outputfile=/var/reana/users/00000000-0000-0000-0000-000000000000/workflows/1bb4b9de-3d77-452e-8d59-c49be4086293/helloworld/greetings.txt sleeptime=0.0


Completed
```

With `adage==0.10.1` (and `networkx==1.11`) everything runs as usual.

Adage release: https://github.com/yadage/adage/releases/tag/v0.10.2

### LINKED ISSUE #221: Add `python_requires=">=3.8"`
For similar motivations as https://github.com/reanahub/reana-client/pull/581 add `python_requires=">=3.8"` (given testing is only happening on Python 3.8 atm) to [`setup.py`](https://github.com/reanahub/reana-workflow-engine-yadage/blob/ac194f6d2972ea77642ba5d3f10d2da2173fff0e/setup.py#L66).

### PR DESCRIPTION
**Pull request by mvidalgarcia**
closes #220
closes #221

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -4,56 +4,56 @@
...
-urllib3==1.26.7           # via requests
+swagger-spec-validator==2.7.4  # via bravado-core
+typing-extensions==4.0.1  # via bravado
+urllib3==1.26.8           # via requests
 vine==1.3.0               # via amqp
 webcolors==1.11.1         # via jsonschema
 werkzeug==2.0.2           # via reana-commons
-yadage-schemas==0.10.6    # via packtivity, reana-commons, yadage
+yadage-schemas==0.10.7    # via packtivity, reana-commons, yadage
 yadage==0.20.1            # via reana-commons
```
Comment (matthewfeickert) [inline requirements.txt:57]: Sorry to ask for yet another rebuild of the `requirements.txt`, but can you rebuild it with `yadage==0.20.2` given https://github.com/reanahub/reana-commons/pull/333?

Comment (matthewfeickert) [reply requirements.txt]: Thanks!

Comment (mvidalgarcia) [reply requirements.txt]: @matthewfeickert After upgrading, Yadage workflows are failing again:

``` console
$ reana-client logs -w helloworld-yadage-kubernetes     
==> Workflow engine logs
Workflow exited unexpectedly.
stat: path should be string, bytes, os.PathLike or integer, not NoneType
```

Do you have any clue?

Comment (matthewfeickert) [reply requirements.txt]: * First: Very sorry &mdash; This is not good and not the experience we want the REANA team to have when interacting with yadage stuff. :(
* Second: This is only happening after going from `yadage` `v0.20.1` to `yadage` `v0.20.2`, correct? If so, there are a [small number of things that happened between them](https://github.com/yadage/yadage/releases/tag/v0.20.2), and probably could only either be https://github.com/yadage/yadage/pull/92 or https://github.com/yadage/yadage/pull/111.

Can you tell me:
* What installed packaged changed beyond `yadage`, if any?
* If there are any more detail logs?
* If there is a minimal reproducible example that I can run myself for faster debugging purposes?

Comment (mvidalgarcia) [reply requirements.txt]: It works as expected with these changes

```diff
diff --git a/requirements.txt b/requirements.txt
index c20ed6b..31b6cc3 100644
--- a/requirements.txt
+++ b/requirements.txt
@@ -41,7 +41,7 @@ pyrsistent==0.18.0        # via jsonschema
 python-dateutil==2.8.2    # via bravado, bravado-core
 pytz==2021.3              # via bravado-core, fs
 pyyaml==5.4.1             # via bravado, bravado-core, packtivity, reana-commons, yadage, yadage-schemas
-reana-commons[yadage]==0.8.3  # via reana-workflow-engine-yadage (setup.py)
+reana-commons[yadage]==0.8.3a1  # via reana-workflow-engine-yadage (setup.py)
 requests[security]==2.26.0  # via bravado, packtivity, reana-workflow-engine-yadage (setup.py), yadage, yadage-schemas
 rfc3987==1.3.8            # via jsonschema, reana-workflow-engine-yadage (setup.py)
 simplejson==3.17.5        # via bravado, bravado-core
@@ -54,7 +54,7 @@ vine==1.3.0               # via amqp
 webcolors==1.11.1         # via jsonschema
 werkzeug==2.0.2           # via reana-commons
 yadage-schemas==0.10.6    # via packtivity, reana-commons, yadage
-yadage==0.20.2            # via reana-commons
+yadage==0.20.1            # via reana-commons
 
 # The following packages are considered to be unsafe in a requirements file:
 # setuptools
diff --git a/setup.py b/setup.py
index 7933c33..505c852 100644
--- a/setup.py
+++ b/setup.py
@@ -47,7 +47,7 @@ install_requires = [
     "mock>=3.0,<4",
     "pydot>=1.0.29",  # FIXME needed only if yadage visuale=True.
     "pydotplus>=2.0.2",  # FIXME needed only if yadage visuale=True.
-    "reana-commons[yadage]>=0.8.3,<0.9.0",
+    "reana-commons[yadage]==0.8.3a1",
     "requests>=2.25.1",
     "rfc3987==1.3.8",  # FIXME remove once yadage-schemas solves yadage deps.
 ]
```

Is there anything we can do to fix it?

Comment (matthewfeickert) [reply requirements.txt]: > It works as expected with these changes:

Is the 

```diff
-reana-commons[yadage]==0.8.3  # via reana-workflow-engine-yadage (setup.py)
+reana-commons[yadage]==0.8.3a1  # via reana-workflow-engine-yadage (setup.py)
```

relevant at all? Or is that just an artifact of the diff? Do things work with `reana-commons==0.8.3a1` and `yadage==0.20.1`?

> Is there anything we can do to fix it?

I can yank the release if we aren't able to understand what is happening and then have you roll back the relevant part of https://github.com/reanahub/reana-commons/pull/333, but I think the fastest way forward is if there is a way for me to be able to get an example or an environment where I can test this and see what's actually happening in the logs.

Comment (lukasheinrich) [reply requirements.txt]: hi @mvidalgarcia is there some fairly isolated engine-specific test suite / setup that would allow @matthewfeickert do test the setup with the new releases short of deploying a full REANA cluster?

**Review by audrium — APPROVED** [review]
LGTM! Temporary pinning `reana-commons[yadage]==0.8.3a1` solves the issue

**Comment by tiborsimko** [comment]
FWIW pinning alpha package `r-commons 0.8.3a1` in non-alpha `r-w-e-yadage 0.8.1` does not seem very good, because it leads to conflicts:

```console
$ docker build -t reanahub/reana-workflow-engine-yadage:latest .
---> Running in 29ffcf96a0ac
reana-workflow-engine-yadage 0.8.1 has requirement reana-commons[yadage]==0.8.3a1, but you have reana-commons 0.8.3.
The command '/bin/sh -c pip check' returned a non-zero code: 1
```

when one  has `r-commons 0.8.3` checked out locally for `r-client 0.8.1`  PR...

File: requirements.txt (hunk 1 above)
Comment (tiborsimko) [reply requirements.txt]: I think we have basically three options:

**A. Enrich Yadage's own test suite to have all the functions that REANA uses covered and tested there.** This may be tricky, since `r-w-e-yadage` uses a lot of functions from `yadage`, `adage`, `yadageschemas`, and `packtivity`, and some of them may be perhaps considered Yadage's "internal functions" and not Yadage's "API-like" functions in purpose. Examples include: `steering_ctx`, `dagstate`, `finalize_inputs`, and more. We'd have to go carefully through `r-w-e-yadage` integration code and make a list and make sure that all these usages are well covered in Yadage's own internal test suite. This would take time, and may be fragile if either Yadage implementation code changes or REANA changes code without updating one another.

**B. Deploy local REANA cluster with new Yadage versions via Kind and test there.** This is the fully opposite option, not to test in Yadage but to test in REANA. This is more natural, and we do have `reana-dev` helper script which helps in setting things up. We could further build a tiny shell script wrapper to make it even easier to automatise. However it would require running Kind etc, so it would bring some REANA development environment to learn, which you were seeking to avoid.

**C. Release constraint on REANA client and REANA server pinning the same Yadage versions.** This would allow incrementing Yadage versions more freely between client and cluster. Which in my eyes would be probably the best solution, since it would both (i) simplify option B and (ii) allow for using any local Yadage version that the user may have installed, whilst still connecting to any compatible remote REANA's Yadage version supported by the cluster. I think this should mostly work, there is no reason why two different Yadage 0.20.x wouldn't be able to run the same workflow definition!
The option C, besides relaxing unnecessary Yadage version constraint to the users on the client side, would allow us to implement client in a faster language (e.g. Go) whilst moving all the advanced workflow validation functionality to the server-side (in order to run on the exactly the same Yadage version that the remote cluster supports).

TL;DR The option A moves all testing to Yadage side, the option B moves all testing to the REANA side, the option C tries to dissociate Yadage clients from REANA Yadage server, so that both can keep evolving independently of each other.

Comment (matthewfeickert) [reply requirements.txt]: @tiborsimko Thanks for the summary.

The `yadage` org libraries have bad codecoverage at the moment. I don't think that's going to drastically improve quickly as @lukasheinrich and I are both more "keeping things working" maintainers at the moment rather than active developers. So while GitHub Issues that contain the information of

> go carefully through r-w-e-yadage integration code and make a list and make sure that all these usages are well covered in Yadage's own internal test suite.

would be welcome and useful that's a lot of time and work on the REANA team side of things.

> **Deploy local REANA cluster with new Yadage versions via Kind and test there**

I like this idea in general as it seems to be the natural relationship between an application (REANA) and libraries (everything under the `yadage` org) of testing any library before release and deployment.

I don't have any thoughts on your suggested Option C as that seems to be a decision for the REANA team in general, but yes, I agree with the sentiment that

> there is no reason why two different Yadage 0.20.x wouldn't be able to run the same workflow definition!
